node/internal/actors/external_agent.go

```diff
@@ -34,16 +52,77 @@ func (a *ExternalAgent) PostStop(ctx context.Context) error {
 }
 
 func (a *ExternalAgent) Receive(ctx *goakt.ReceiveContext) {
-	switch ctx.Message().(type) {
+	switch msg := ctx.Message().(type) {
 	case *goaktpb.PostStart:
+		if err := a.startBinary(); err != nil {
+			ctx.Err(errors.Join(errors.New("Failed to load and start agent binary"), err))
+			return
+		}
+		a.self = ctx.Self()
+		_ = a.self.ActorSystem().ScheduleOnce(
+			context.Background(),
+			&actorproto.CheckRegistered{},
+			a.self,
+			registrationTimeout)
 		a.logger.Info("External agent for workload type is running", slog.String("name", ctx.Self().Name()))
+	case *actorproto.AgentRegistered:
+		a.createNatsConnection(msg.InternalNatsUrl, msg.InternalNkey, msg.InternalNkeySeed)
+		ctx.Become(a.RegisteredAgentReceive)
+	case *actorproto.CheckRegistered:
+		a.logger.Error("Agent was not registered within timeout period. Agent actor terminating", slog.String("agent", a.agentOptions.Name))
+		ctx.Stop(a.self)
+	default:
+		ctx.Unhandled()
+	}
+}
+
+func (a *ExternalAgent) RegisteredAgentReceive(ctx *goakt.ReceiveContext) {
+	switch ctx.Message().(type) {
 	case *actorproto.QueryWorkloads:
 		a.queryWorkloads(ctx)
 	default:
 		ctx.Unhandled()
 	}
 }
 
+func (a *ExternalAgent) startBinary() error {
+	// TODO: download the artifact, create an OsProcess, run it
+
+	u, err := url.Parse(a.internalNatsUrl)
+	if err != nil {
+		return err
+	}
+	seed, err := a.internalNatsCreds.nkey.Seed()
+	if err != nil {
+		return err
+	}
+	env := make(map[string]string)
+	env[agentcommon.EnvNatsHost] = u.Host
+	env[agentcommon.EnvNatsPort] = u.Port()
+	env[agentcommon.EnvNatsNkey] = string(seed)
+
+	// TODO: Note that host services information needs to be passed to the agent in start
+	// workload request because the Hs connections are per-workload
+
+	return nil
+}
+
+func (a *ExternalAgent) createNatsConnection(url string, _ string, seed string) {
+	var err error
+	opt, err := nats.NkeyOptionFromSeed(seed)
+	if err != nil {
+		a.logger.Error("Failed to extract an nkey option from the nkey seed", slog.Any("error", err))
+		return
+	}
+	a.conn, err = nats.Connect(url, opt)
```

Comment:
can we add the "name" option here so the connections are distinguished more easily

Reply:
Sure but it's a connection to a server no one knows about nor can they connect to it. 

Will add the name tomorrow